File: server/app/schemas/contracts.py
"""Core data contracts for Black Swan FP&A agents (M, Q, Eve, 007).

Strict Pydantic V2 models defining the structured inter-agent data flow.
"""

from typing import Any, Dict, List, Literal
from pydantic import BaseModel, ConfigDict, Field


# -----------------------------------------------------------------------------
# Sub-models & Supporting Types
# -----------------------------------------------------------------------------

class Subtask(BaseModel):
    """Sub-task allocated by M Orchestrator."""
    model_config = ConfigDict(extra="allow")

    step_id: str
    agent_assigned: str
    goal: str


class AnomalyDetected(BaseModel):
    """Diagnostic variance anomaly detected by Q."""
    model_config = ConfigDict(extra="allow")

    field: str
    expected: Any
    actual: Any
    delta_pct: float


class ChartSeries(BaseModel):
    """Series specification for Recharts visualization."""
    model_config = ConfigDict(extra="allow")

    key: str
    label: str
    color_role: str


class ChartSpec(BaseModel):
    """Chart specification produced by Eve."""
    model_config = ConfigDict(extra="allow")

    chart_type: Literal["line", "bar", "area", "waterfall"]
    x_axis_key: str
    series: List[Dict[str, str]]
    title: str


class FormulaLedgerEntry(BaseModel):
    """Plain-language formula verification ledger entry."""
    model_config = ConfigDict(extra="allow")

    metric: str
    formula: str
    computation_step: str


class SensitivityLever(BaseModel):
    """Dynamic what-if sensitivity slider parameters for Agent007."""
    model_config = ConfigDict(extra="allow")

    id: str
    label: str
    min_val: float
    max_val: float
    default_val: float
    unit: str
    step: float


# -----------------------------------------------------------------------------
# Agent Primary Contracts
# -----------------------------------------------------------------------------

class M_Plan(BaseModel):
    """Orchestrator plan: decomposes query, allocates subtasks, and tracks execution status."""
    model_config = ConfigDict(extra="allow")

    task_id: str
    user_query: str
    subtasks: List[Dict[str, str]] = Field(
        ...,
        description="Subtasks containing step_id, agent_assigned, and goal",
    )
    status: Literal["planning", "executing", "completed", "failed"]


class Q_Diagnostic(BaseModel):
    """Data & Diagnostics output: executed SQL, tabular output, and variance anomalies."""
    model_config = ConfigDict(extra="allow")

    table_headers: List[str]
    rows: List[Dict[str, Any]]
    summary_findings: List[str]
    executed_sql: str
    execution_time_ms: float
    anomalies_detected: List[Dict[str, Any]] = Field(
        ...,
        description="Anomalies containing field, expected, actual, delta_pct",
    )


class Eve_Audit(BaseModel):
    """Audit & Visualization output: chart specs, formula receipts, assumptions, and confidence."""
    model_config = ConfigDict(extra="allow")

    chart_spec: ChartSpec
    formula_ledger: List[Dict[str, str]] = Field(
        ...,
        description="Formula ledger entries with metric, formula, and computation_step",
    )
    assumptions: List[str]
    confidence_score: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Confidence score bounded between 0.0 and 1.0",
    )
    plain_language_narrative: str


class Agent007_Strategy(BaseModel):
    """Strategy output: executive actions and bounded what-if sensitivity levers."""
    model_config = ConfigDict(extra="allow")

    headline_recommendation: str
    strategic_actions: List[str]
    estimated_impact: str
    sensitivity_levers: List[Dict[str, Any]] = Field(
        ...,
        description="Bounded levers containing id, label, min_val, max_val, default_val, unit, step",
    )

File: web/src/types/contracts.ts
/**
 * Core data contracts for Black Swan FP&A agents (M, Q, Eve, 007).
 * Strict 1:1 TypeScript parity with Python Pydantic V2 models.
 */

// -----------------------------------------------------------------------------
// Sub-types & Supporting Interfaces
// -----------------------------------------------------------------------------

export interface Subtask {
  step_id: string;
  agent_assigned: string;
  goal: string;
  [key: string]: string;
}

export interface AnomalyDetected {
  field: string;
  expected: any;
  actual: any;
  delta_pct: number;
  [key: string]: any;
}

export interface ChartSeries {
  key: string;
  label: string;
  color_role: string;
  [key: string]: string;
}

export type ChartType = "line" | "bar" | "area" | "waterfall";

export interface ChartSpec {
  chart_type: ChartType;
  x_axis_key: string;
  series: Array<ChartSeries | Record<string, string>>;
  title: string;
  [key: string]: any;
}

export interface FormulaLedgerEntry {
  metric: string;
  formula: string;
  computation_step: string;
  [key: string]: string;
}

export interface SensitivityLever {
  id: string;
  label: string;
  min_val: number;
  max_val: number;
  default_val: number;
  unit: string;
  step: number;
  [key: string]: any;
}

// -----------------------------------------------------------------------------
// Agent Primary Contracts
// -----------------------------------------------------------------------------

export type M_PlanStatus = "planning" | "executing" | "completed" | "failed";

export interface M_Plan {
  task_id: string;
  user_query: string;
  intent?: string;
  hypotheses?: string[];
  required_metrics?: string[];
  sql_objective?: string;
  strategic_focus?: string;
  subtasks: Array<Record<string, string>>;
  status: M_PlanStatus;
  [key: string]: any;
}

export interface Q_Diagnostic {
  table_headers: string[];
  rows: Array<Record<string, any>>;
  summary_findings: string[];
  executed_sql: string;
  execution_time_ms: number;
  anomalies_detected: Array<Record<string, any>>;
  [key: string]: any;
}

export interface Eve_Audit {
  chart_spec: {
    chart_type: ChartType;
    x_axis_key: string;
    series: Array<Record<string, string>>;
    title: string;
    [key: string]: any;
  };
  formula_ledger: Array<Record<string, string>>;
  assumptions: string[];
  confidence_score: number; // 0.0 to 1.0
  plain_language_narrative: string;
  [key: string]: any;
}

/**
 * NAMING NOTE: Agent007_Strategy is deliberately named this way (not 007_Strategy)
 * because TypeScript and Python identifiers cannot start with a digit.
 */
export interface Agent007_Strategy {
  headline_recommendation: string;
  strategic_actions: string[];
  estimated_impact: string;
  sensitivity_levers: Array<Record<string, any>>;
  [key: string]: any;
}

// -----------------------------------------------------------------------------
// SSE Stream Event & Report Block
// -----------------------------------------------------------------------------

export type SSEEventType =
  | "m_plan"
  | "q_diagnostic"
  | "eve_audit"
  | "007_strategy"
  | "status_update"
  | "out_of_scope"
  | "pipeline_complete"
  | "error";

export type AgentRole = "M" | "Q" | "Eve" | "007" | "System";

export interface SSEStreamEvent {
  event_type: SSEEventType;
  agent: AgentRole;
  payload: Record<string, any>;
  timestamp: string;
}

export interface ReportBlock {
  id: string;
  query: string;
  timestamp: string;
  mPlan: M_Plan | null;
  qDiagnostic: Q_Diagnostic | null;
  eveAudit: Eve_Audit | null;
  strategy007: Agent007_Strategy | null;
  outOfScope?: {
    refusal_message: string;
    suggested_queries: string[];
  } | null;
  quotaExceeded?: boolean;
}
